tests/test_nanopub.py

```diff
@@ -0,0 +1,204 @@
+import pytest
+from rdflib import BNode, Graph, Literal, URIRef
+
+from nanopub import Nanopub, NanopubClaim, NanopubConf, NanopubRetract, NanopubUpdate, create_nanopub_index, namespaces
+from nanopub.templates.nanopub_introduction import NanopubIntroduction
+from tests.conftest import default_conf, java_wrap, skip_if_nanopub_server_unavailable
+
+
+def test_nanopub_sign_uri():
+    expected_np_uri = "http://purl.org/np/RAoXkQkJe_lpMhYW61Y9mqWDHa5MAj1o4pWIiYLmAzY50"
+    assertion = Graph()
+    assertion.add((
+        URIRef('http://test'), namespaces.HYCL.claims, Literal('This is a test of nanopub-python')
+    ))
+    np = Nanopub(
+        conf=default_conf,
+        assertion=assertion
+    )
+    java_np = java_wrap.sign(np)
+    np.sign()
+    assert np.has_valid_signature
+    assert np.source_uri == expected_np_uri
+    assert np.source_uri == java_np
+
+
+
+def test_nanopub_sign_uri2():
+    expected_np_uri = "http://purl.org/np/RAoXkQkJe_lpMhYW61Y9mqWDHa5MAj1o4pWIiYLmAzY50"
+    np = Nanopub(
+        conf=default_conf,
+    )
+    np.assertion.add((
+        URIRef('http://test'), namespaces.HYCL.claims, Literal('This is a test of nanopub-python')
+    ))
+    java_np = java_wrap.sign(np)
+    np.sign()
+    assert np.has_valid_signature
+    assert np.source_uri == expected_np_uri
+    assert np.source_uri == java_np
+
+
+def test_nanopub_sign_bnode():
+    expected_np_uri = "http://purl.org/np/RAclARDMZxQ0yLKu3enKS4-CGubi2coQUvyb7BXF3XRvY"
+    assertion = Graph()
+    assertion.add((
+        BNode('test'), namespaces.HYCL.claims, Literal('This is a test of nanopub-python')
+    ))
+    np = Nanopub(
+        conf=default_conf,
+        assertion=assertion
+    )
+    np.sign()
+    assert np.has_valid_signature
+    assert np.source_uri == expected_np_uri
+
+
+def test_nanopub_sign_bnode2():
+    expected_np_uri = "http://purl.org/np/RA2bruKoZi0snNNfQCkB2qvhCnscTt9Wmz2_rSGnwB2nQ"
+    assertion = Graph()
+    assertion.add((
+        BNode('test'), namespaces.HYCL.claims, Literal('This is a test of nanopub-python')
+    ))
+    assertion.add((
+        BNode('test2'), namespaces.HYCL.claims, Literal('This is another test of nanopub-python')
+    ))
+    np = Nanopub(
+        conf=default_conf,
+        assertion=assertion
+    )
+    np.sign()
+    assert np.source_uri == expected_np_uri
+
+
+def test_nanopub_publish():
+    expected_np_uri = "http://purl.org/np/RAoXkQkJe_lpMhYW61Y9mqWDHa5MAj1o4pWIiYLmAzY50"
+    assertion = Graph()
+    assertion.add((
+        URIRef('http://test'), namespaces.HYCL.claims, Literal('This is a test of nanopub-python')
+    ))
+    np = Nanopub(
+        conf=default_conf,
+        assertion=assertion
+    )
+    java_np = java_wrap.sign(np)
+    np.publish()
```

Comment:
So is this still publishing it to the actual test server? (or some local mock?)

Does it not risk being flaky, since the server isn't always 100% responsive. I think previously we allowed 10 retries or something. Maybe you don't need that anymore? (sorry if I've misunderstood something)